internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/util"
+	"golang.org/x/sync/errgroup"
+)
+
+// downloadRangeToFile is a helper function to download a given range of object
+// from GCS into given file handle.
+//
+// This function doesn't take locks and can be executed parallely.
+func (job *Job) downloadRangeToFile(ctx context.Context, cacheFile *os.File, start, end int64) error {
+	newReader, err := job.bucket.NewReader(
+		ctx,
+		&gcs.ReadObjectRequest{
+			Name:       job.object.Name,
+			Generation: job.object.Generation,
+			Range: &gcs.ByteRange{
+				Start: uint64(start),
+				Limit: uint64(end),
+			},
+			ReadCompressed: job.object.HasContentEncodingGzip(),
+		})
+	if err != nil {
+		err = fmt.Errorf("downloadRangeToFile: error in creating NewReader with start %d and limit %d: %w", start, end, err)
+		return err
+	}
+	defer func() {
+		closeErr := newReader.Close()
+		if closeErr != nil {
+			logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, closeErr)
+		}
+	}()
+
+	monitor.CaptureGCSReadMetrics(ctx, util.Parallel, end-start)
+
+	// Copy the contents from NewReader to cache file at appropriate offset.
+	offsetWriter := io.NewOffsetWriter(cacheFile, start)
+	_, err = io.CopyN(offsetWriter, newReader, end-start)
+	if err != nil {
+		err = fmt.Errorf("downloadRangeToFile: error at the time of copying content to cache file %w", err)
+	}
+	return err
+}
+
+// parallelDownloadObjectAsync does parallel download of the backing GCS object
+// into a file as part of file cache using multiple NewReader method of
+// gcs.Bucket running in parallel
+//
+// Note: There can only be one async parallel or non-parallel download running
+// for a job at a time.
+// Acquires and releases LOCK(job.mu)
+func (job *Job) parallelDownloadObjectAsync() {
+	// Cleanup the async job in all cases - completion/failure/invalidation.
+	defer job.cleanUpDownloadAsyncJob()
+
+	// Create, and open cache file for writing object into it.
+	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_RDWR)
+	if err != nil {
+		err = fmt.Errorf("parallelDownloadObjectAsync: error in creating cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	defer func() {
+		err = cacheFile.Close()
+		if err != nil {
+			err = fmt.Errorf("parallelDownloadObjectAsync: error while closing cache file: %w", err)
+			job.failWhileDownloading(err)
+		}
+	}()
+
+	notifyInvalid := func() {
+		job.mu.Lock()
+		job.status.Name = Invalid
+		job.notifySubscribers()
+		job.mu.Unlock()
+	}
+
+	var start, end int64
+	end = int64(job.object.Size)
+	var parallelReadRequestSize = int64(job.fileCacheConfig.ReadRequestSizeMB) * cacheutil.MiB
+	var maxSingleStepReadSize = parallelReadRequestSize * int64(job.fileCacheConfig.DownloadParallelismPerFile)
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
```

Comment:
Looking at the code more i see only the default clause is different for sequential download vs parallel download, why can;t we put a if -else based on flag and invoke seperate functions?

Reply:
done. PTAL